### ISSUE TITLE
Lua - input cannot be swapped between joypads using joypad.set() and joypad.get()

### ISSUE THREAD

**Issue by AndreaJens** [TASVideos/BizHawk#2656]
[//]: # "This issue description supports Markdown syntax (this is what comments look like). There's a cheatsheet here: https://guides.github.com/features/mastering-markdown/"
[//]: # "You can leave these comments here or delete them. Also, please remember to search for similar issues before writing anything, including in closed issues!"
[//]: # "One more thing: if you're on Linux, please comment on #1430 instead of opening an issue so we don't annoy the other devs."

### Summary
[//]: # "Briefly describe what's broken. Include relevant details: loaded core, loaded rom's hash, open tools, running scripts... You can embed a screenshot if it's easier to show the bug, but if you need more than one please put them at the end."
When using the Lua scripting on the PSX emulator, the following code doesn't seem to work:
```
-- shallow copy function for tables --
	function shallowcopy(inputTable)
		local outputTable = {}
		for key, value in pairs(inputTable) do
			outputTable[key] = value
		end
		return outputTable
	end
	
-- main routine
	while true do
		inputTableP1 = shallowcopy(joypad.get(1))
		inputTableP2 = shallowcopy(joypad.get(2))
		joypad.set(inputTableP1, 2)
		joypad.set(inputTableP2, 1)
		emu.frameadvance()
	end
```
#### Expected outcome
The input of joypad 1 and joypad 2 joypad should be swapped (joypad 1 would control player 2's character and vice versa). The function called shallowcopy should prevent any reference to be overwritten or aliased.

#### Observed outcome
* No input is seemingly received by the game, except for the analog sticks;
* Printing the output before `joypad.set()` return always a table with all inputs set to false;
* Removing either of the ` joypad.set()` functions just causes the input to be duplicated between the joypads.

### Repro
[//]: # "If you can't figure out the list of steps, delete this section and put 'heisenbug' in the summary somewhere. If a Lua script can cause the bug, you can embed that instead (as simple as possible please)."
1. load the above script into the Lua console while playing a PSX game (tested with Digimon Rumble Arena (US-NTSC))
2. no input is received by either joypads connected.

### Host env.
[//]: # "List the computers you've found the bug with. If there's a version that doesn't have the bug, please put that in too. Here are some examples:"
- BizHawk 2.5.2; Win10 Home; processor: Intel i7, 8th Gen; graphic card: NVidia RTX2070

[//]: # "(screenshots, if applicable)"

[//]: # "That's it! If you'd like to help more, you could try a dev build (see Testing in the readme) or an older release. Click submit now and you can edit it later."

**Comment by zeromus** [comment]
print the names of the buttons. you will probably see that they have a player number baked into them. I know this isn't what you expect.
the joypad.get and set apis should probably not have player numbers on them, for this reason.

**Comment by AndreaJens** [comment]
Hi, zeromus.
I have printed the output. No player information is stored inside:

```
console.log("Output")
console.log(inputTableP1)
console.log(inputTableP2)
```

Output
"Circle": "False"
"Cross": "False"
"Down": "False"
"L1": "False"
"L2": "False"
"L3": "False"
"Left": "False"
"LStick X": "128"
"LStick Y": "128"
"MODE": "False"
"R1": "False"
"R2": "False"
"R3": "False"
"Right": "False"
"RStick X": "128"
"RStick Y": "128"
"Select": "False"
"Square": "False"
"Start": "False"
"Triangle": "False"
"Up": "False"

"Circle": "False"
"Cross": "False"
"Down": "False"
"L1": "False"
"L2": "False"
"L3": "False"
"Left": "False"
"LStick X": "128"
"LStick Y": "128"
"MODE": "False"
"R1": "False"
"R2": "False"
"R3": "False"
"Right": "False"
"RStick X": "128"
"RStick Y": "128"
"Select": "False"
"Square": "False"
"Start": "False"
"Triangle": "False"
"Up": "False"

**Comment by zeromus** [comment]
huh, guess im wrong
try joypad.getimmediate instead of joypad.get

**Comment by AndreaJens** [comment]
Thanks for the suggestion. Unfortunately, I've already done it, it doesn't change anything.
Might it be a problem with Octoshock (i.e. the receiving end of the lua commands)?

**Comment by AndreaJens** [comment]
Another thing I observed, that might maybe help: 
If I modify the code like this:
```
-- main routine
	while true do
		inputTableP1 = shallowcopy(joypad.get(1))
		inputTableP2 = shallowcopy(joypad.get(2))
		joypad.set(inputTableP2, 1)
		joypad.set(inputTableP1, 1)
		emu.frameadvance()
	end
```
#### Expected 
Everything should work as normal, since I restored the input after changing it
####Observed 
The first joypad doesn't produce any input for the game

**Comment by YoshiRulz** [comment]
Use `input.get()` with the keybinds unset.
edit: per https://github.com/TASEmulators/BizHawk/issues/4575#issuecomment-3620309486

**Comment by AndreaJens** [comment]
Hi, YoshiRulz, thanks for the reply. What does it mean exactly? Where should I put input.get() in the above script?

**Comment by AndreaJens** [comment]
Yeah, I have to ask to reopen the ticket: `input.get()` doesn't solve the issue, as it returns only the raw input, instead of the emulator-specific buttons for a specific joypad. Who should I ask for it?

**Comment by YoshiRulz** [comment]
We don't have a mechanism for swapping players. To hack one in with Lua, as you seem to be trying, you'll need to circumvent the keybind system entirely: read with `input.get`, map keybinds, and then write with `joypad.set`/`joypad.setanalog`.

**Comment by zeromus** [comment]
he's programming the mechanism for swapping players. this should work.
my understanding was that he was gonna need to clear the set input in order to read it again fresh from the input, but JoypadIntersection.lua doesn't seem to work right now

**Comment by zeromus** [comment]
I believe you broke it in 1042f746f4b2d470b72fb9d98243d72ff04f44be when you changed logic that supported values other than true/false. values other than true/false are meant to remove the input from the overriding. JoypadIntersection.lua should work.

**Comment by AndreaJens** [comment]
> @AndreaJens once yoshirulz fixes that, you need to study the examples instead of trusting your intuition. Our lua implementation doesn't make sense. It's barely limping along nonsense.

Thanks for the reply. Yeah, my problem was that I didn't find examples to try and do what I needed. I'm coding a training mode script for a PS1 fighting game and bumped into this issue when coding the "record input for dummy" function and basically needed to swap the inputs of joypad 1 and 2 to prevent player 1 from moving while recording player 2. Until this point, my intuition worked fine and the Lua implementation did its job wonderfully, I must say :'D

**Comment by zeromus** [comment]
You didn't know you needed to clear the input. There's no reason to expect that, really. I didnt see any docs explaining it. You just have to "know" thats how it (was supposed to) work. Now that you do know that, there are several examples (which will work once the bug is fixed)

**Comment by YoshiRulz** [comment]
The current JoypadApi should still let you unset by omitting the entry (in Lua, this is equivalent to setting it to `nil`). The overload which takes a dict calls the other, which handles setting `null`:
https://github.com/TASEmulators/BizHawk/blob/3fb35b159982fb7fbd9e8e98e65a9bc601cdeffe/src/BizHawk.Client.Common/Api/Classes/JoypadApi.cs#L54-L77

**Comment by AndreaJens** [comment]
Once again, thanks for all help. Where can I actually find tutorials for learning more? I apologize, but I really didn't understand where to look for them.
Also, I might have misunderstood, but for clearing the buffer I tried both
```
	while true do
		inputTableP1 = shallowcopy(joypad.get(1))
		inputTableP2 = shallowcopy(joypad.get(2))
		joypad.set(inputTableP1, 2)
		joypad.set(inputTableP2, 1)
		emu.frameadvance()
		joypad.set({Square = nil, Triangle = nil, Cross = nil, Circle = nil, Up = nil,
					Down = nil, Left = nil, Right = nil, L1 = nil, R1 = nil}, 2)
		joypad.set({Square = nil, Triangle = nil, Cross = nil, Circle = nil, Up = nil,
					Down = nil, Left = nil, Right = nil, L1 = nil, R1 = nil}, 1)
	end
```
and

```
	while true do
		inputTableP1 = shallowcopy(joypad.get(1))
		inputTableP2 = shallowcopy(joypad.get(2))
		joypad.set(inputTableP1, 2)
		joypad.set(inputTableP2, 1)
		emu.frameadvance()
		input.get()
	end
```

But this didn't help, the behaviour is still the same as before changing stuff. Does it mean that there is currently no workaround outside of waiting for a fix? 
If not, what is the issue here with this script and where do I have to clear the input?
Apparently, after I assign the new tables, no input seems to be processed when `emu.frameadvance()` is called, no matter what I do in-between.
Thanks in advance and sorry for the mess.

**Comment by zeromus** [comment]
@YoshiRulz ok that's a fine design, but you broke the demos. please fix the demos or fix the code so nonsense keys continue to be the same as removed. 
@AndreaJens bizhawk came with examples
@ everyone - I fixed an issue that made this particular pattern of joypad get and set malfunction

**Comment by AndreaJens** [comment]
@zeromus I have checked every example script inside the LUA folder and tried to replicate the clearing step, but it still doesn't work.

```
-- shallow copy function for tables --
	function shallowcopy(inputTable)
		local outputTable = {}
		for key, value in pairs(inputTable) do
			outputTable[key] = value
		end
		return outputTable
	end
	
	local turnoff = {Triangle = null, Square = null, Circle = null, Cross = null,
			   Up = null, Down = null, Right = null, Left = null, L1 = null,
			   R1 = null, Start = null, L2 = null, R2 = null, Select = null, Start = null}
    local inputTableP1 = {}
    local inputTableP2 = {}
	
-- main routine
	while true do
		-- joypad.set(turnoff, 1)
		-- joypad.set(turnoff, 2)
		inputTableP1 = shallowcopy(joypad.get(1))
		inputTableP2 = shallowcopy(joypad.get(2))
		-- joypad.set(turnoff, 1)
		-- joypad.set(turnoff, 2)
		joypad.set(inputTableP1, 2)
		joypad.set(inputTableP2, 1)
		emu.frameadvance()
		-- joypad.set(turnoff, 1)
		-- joypad.set(turnoff, 2)
	end
```
The commented parts are where I tried to use the clearing function, copied almost verbatim from the examples and adapted to the PSX controller layout. No matter where or when I clear the input, the joypads still yield no input when swapped.

**Comment by YoshiRulz** [comment]
@AndreaJens You are *still* using `joypad.get` instead of `input.get`, with no keybind lookup table in sight. At this point I'm about to get a Windows machine and write the damn thing for you, when I should be working on UI.
Bonus criticism: There's no need for cloning the tables you get from API calls. Lua uses `nil`, not `null` (the latter is an undefined global which happens to evaluate to `nil`). And as I mentioned above, `{ ["key"] = nil }` is equivalent to omitting the entry.

**Comment by AndreaJens** [comment]
@YoshiRulz no need to be rude. I have been told two contrasting things, and now I'm legitimately confused. 

From the replies above I got that I *still could* use `joypad.get()/joypad.set()`, just by clearing the buffer with {} and I didn't *need* to use` input.get()` at all. If this is not the truth, what did @zeromus's comment meant with *clearing the buffer using joypad.set()?*. Form how it was written, it looked like a one-line replacement.

If it's not, might I ask you which LUA sample script should have I used as a base? I have tried to find an example, since I was directed to browse them, but I cannot find anything about it. The only examples I've found showed the input cleaning part using joypad.set(), which I copied almost verbatim, and didn't work. 

Don't take it personally, but *how* am I supposed to figure it out on my own if: there is no example touching the topic of binding input.get to a lookup table for the a controller, `input.get()` is never used with a lookup table in a sample LUA file in the repository, and I am told *to just clear the input using joypad.set()* one comment above?

About the bonus criticism: I have always used `nil `before, but then I brosed the sample LUA scripts with input clearing, `null `is used (see Joypad_Set.lua and Joypad_withControllerNumbers.lua), so I assumed I was doing something wrong in using `nil` or that `null `was used internally by the interpreter instead.

**Comment by YoshiRulz** [comment]
Here's what I came up with for N64:
```lua
local button_map = {
	["A"] = "A",
--	["B"] = "",
	["X"] = "B",
--	["Y"] = "",
--	["Back"] = "",
	["Start"] = "Start",
	["DpadUp"] = "DPad U",
	["DpadDown"] = "DPad D",
	["DpadLeft"] = "DPad L",
	["DpadRight"] = "DPad R",
	["LeftShoulder"] = "L",
	["RightShoulder"] = "R",
	["LeftTrigger"] = "Z",
--	["RightTrigger"] = "",
	["LStickUp"] = "A Up",
	["LStickDown"] = "A Down",
	["LStickLeft"] = "A Left",
	["LStickRight"] = "A Right",
	["RStickUp"] = "C Up",
	["RStickDown"] = "C Down",
	["RStickLeft"] = "C Left",
	["RStickRight"] = "C Right",
};
local player_map = {
	["1"] = 2,
	["2"] = 1,
--	["3"] = 3,
--	["4"] = 4,
};
set_player_map = function(new_map)
	player_map = new_map;
end
local set_vinput = function()
	local vinput = {};
	for k, v in pairs(player_map) do
		vinput[v] = {};
	end
	local host_input = input.get();
	for k, v in pairs(host_input) do
		if (string.sub(k, 1, 1) == "X") then
			local i = string.find(k, " ");
			local player = player_map[string.sub(k, 2, i - 1)];
			local button = button_map[string.sub(k, i + 1, #k)];
			if (button ~= nil) then
				vinput[player][button] = v;
			end
		end
	end
	for k, v in pairs(vinput) do
		joypad.set(v, k);
	end
end
while true do
	set_vinput();
	emu.frameadvance();
end
```

**Comment by zeromus** [comment]
and here's mine when you do it the way it was all intended to work

```
-- main routine
while true do
	local inputTableP1 = joypad.get(1)
	local inputTableP2 = joypad.get(2)
	joypad.set(inputTableP1, 2)
	joypad.set(inputTableP2, 1)
	emu.frameadvance()
	joypad.set({}, 2)
	joypad.set({}, 1)
end
```

**Comment by AndreaJens** [comment]
@YoshiRulz Thanks. I think I understood the point now. I thought there was a function to fetch the keybinding map
from the emulator itself, I didn't understand that I had to provide the map myself. I will settle for this workaround for the moment.
@zeromus The script you posted should work only after the fix, correct? I have copied it verbatim, used it in the emulator and it still causes no input to be processed. I guess at this point there is no other workaround than `input.get()`.

**Comment by zeromus** [comment]
I fixed it. Use a fixed build.

**Comment by AndreaJens** [comment]
Okay, I will compile the new version! Thanks a lot again!

**Comment by vadosnaprimer** [comment]
No need to compile, the latest build is linked in the repo readme.
